torch/csrc/dynamo/eval_frame.c

```diff
@@ -0,0 +1,578 @@
+    DEBUG_TRACE0("arg error");
+    return NULL;
+  }
+  if (callback != Py_None && callback != Py_False &&
+      !PyCallable_Check(callback)) {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected a callable");
+    return NULL;
+  }
+  DEBUG_TRACE(
+      "python enabled=%d and is run_only=%d",
+      callback != Py_None,
+      callback == Py_False);
+  return set_eval_frame(callback, PyThreadState_GET());
+}
+
+static PyObject* reset_code(PyObject* dummy, PyObject* args) {
+  PyObject* code = NULL;
+  if (!PyArg_ParseTuple(args, "O:code", &code)) {
+    DEBUG_TRACE0("arg error");
+    return NULL;
+  }
+  if (!PyCode_Check(code)) {
+    DEBUG_TRACE0("arg error");
+    PyErr_SetString(PyExc_TypeError, "expected a code object");
+    return NULL;
+  }
+
+  destroy_cache_entry(get_extra((PyCodeObject*)code));
+  set_extra((PyCodeObject*)code, NULL);
+  Py_RETURN_NONE;
+}
+
+static PyObject* unsupported(PyObject* dummy, PyObject* args) {
+  // a dummy C function used in testing
+  PyObject* obj1 = NULL;
+  PyObject* obj2 = NULL;
+  if (!PyArg_ParseTuple(args, "OO", &obj1, &obj2)) {
+    return NULL;
+  }
+  Py_INCREF(obj2);
+  return obj2;
+}
+
+static PyObject* skip_code(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  if (!PyCode_Check(obj)) {
+    PyErr_SetString(PyExc_TypeError, "expected a code object");
+    return NULL;
+  }
+  set_extra((PyCodeObject*)obj, SKIP_CODE);
+  Py_RETURN_NONE;
+}
+
+static PyObject* set_guard_fail_hook(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  Py_XDECREF(guard_fail_hook);
+  if (obj == Py_None) {
+    guard_fail_hook = NULL;
+  } else {
+    guard_fail_hook = obj;
+    Py_INCREF(guard_fail_hook);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyObject* set_guard_error_hook(PyObject* dummy, PyObject* args) {
+  PyObject* obj = NULL;
+  if (!PyArg_ParseTuple(args, "O", &obj)) {
+    return NULL;
+  }
+  Py_XDECREF(guard_error_hook);
+  if (obj == Py_None) {
+    guard_error_hook = NULL;
+  } else {
+    guard_error_hook = obj;
+    Py_INCREF(guard_error_hook);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyMethodDef _methods[] = {
+    {"set_eval_frame", set_eval_frame_py, METH_VARARGS, NULL},
+    {"reset_code", reset_code, METH_VARARGS, NULL},
+    {"unsupported", unsupported, METH_VARARGS, NULL},
+    {"skip_code", skip_code, METH_VARARGS, NULL},
+    {"set_guard_fail_hook", set_guard_fail_hook, METH_VARARGS, NULL},
+    {"set_guard_error_hook", set_guard_error_hook, METH_VARARGS, NULL},
+    {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef _module = {
+    PyModuleDef_HEAD_INIT,
+    "torch._C._dynamo.eval_frame",
+    "Module containing hooks to override eval_frame",
+    -1,
+    _methods};
+
+PyObject* torch_c_dynamo_eval_frame_init(void) {
+  extra_index = _PyEval_RequestCodeExtraIndex(ignored);
```

Comment:
Also shouldn't we pass in `free` as the release functions to properly cleanup cache on code objects?
That won't work with the custom SKIP_CODE I guess, but a simple custom function should do it.